### DISCUSSION TITLE
How to use custom fonts with typst

### DISCUSSION THREAD

**Discussion by royfrancis** [quarto-dev/quarto-cli#7580] [Q&A] [answered]
### Description

I can see the currently available list of fonts using:

```
> quarto typst fonts
DejaVu Sans Mono
Linux Libertine
New Computer Modern
New Computer Modern Math
```

I have a folder called `fonts` in the project directory. And if I run as below, it seems like the new fonts are identified.

```
> quarto typst fonts --font-path fonts
Barlow
DejaVu Sans Mono
Lato
Linux Libertine
New Computer Modern
New Computer Modern Math
```

But it's not saved. Anyway, the custom fonts are not available when running `quarto render index.qmd` with typst format. 

Also tried something like this

```
format:
  typst:
    font-paths: "./fonts"
```
But, didn't seem to do anything much.

Also, is it possible to package fonts into a typst qmd template?

```
> quarto --version
1.4.502
```

**Comment by mcanouil** [comment]
> But it's not saved.

What do you mean? There is nothing to be saved.

> Anyway, the custom fonts are not available when running quarto render index.qmd with typst format.

Do you have any small reproducible example?
Font is set via `set text(font: "Your font", size: 10pt)`, see <https://typst.app/docs/tutorial/advanced-styling/>.

`font-path` is not (at least from the documentation) a supported option, see <https://quarto.org/docs/prerelease/1.4/typst.html#typst-format>.

> Also, is it possible to package fonts into a typst qmd template?

Possibly but not sure there is currently an easy to add the font to the path for Typst to be aware of.

**Reply by royfrancis** [reply]
What I mean is that, after running `quarto typst fonts --font-path ./fonts`, if I run `quarto typst fonts` again, I just get the original list of fonts. It doesn't update the list of available fonts. So, what is this `quarto typst fonts --font-path ./fonts` supposed to do? And how is this argument `--font-path ./fonts` supposed to work with `quarto render`?

**Reply by mcanouil** [reply]
You know `quarto typst` is simply to call the typst version embedded in Quarto, so `--font-path` is a 100% Typst command available in their documentation.
It has basically nothing to do with Quarto render command.
AS with `quarto pandoc`, `quarto typst` is mostly for debugging.

Note that the documentation of Typst CLI shows you can use an environment variable
<img width="702" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/d3b4afd1-0834-4fe4-b2dd-2328e64bc323">
Which you can use in various ways in Quarto, including via `_environment` files, see <https://quarto.org/docs/projects/environment.html>.

This is also described in Typst documentation, see <https://typst.app/docs/reference/text/text/#parameters-font>.

**Reply by royfrancis** [reply]
Hmm.. I have tried setting path `TYPST_FONT_PATHS` to the font directory and then running `quarto render` and I didn't see a difference.

**Reply by royfrancis** [reply]
I am able to change to any of the internal fonts.

````
---
title: Test
format: typst
---

```{=typst}
#set text(font: "New Computer Modern")
```

Here is text in first font.


```{=typst}
#set text(font: "DejaVu Sans Mono")
```

Here is text in second font.

```{=typst}
#set text(font: "Lato")
```

Here is text in external font.
````

![Screenshot 2023-11-14 at 22 49 26](https://github.com/quarto-dev/quarto-cli/assets/5083500/4724bdb1-3ddf-45b3-9c1a-6ab8f75a682b)


But, it's not clear to me how custom/external fonts work.

**Reply by mcanouil** [reply]
Did you use `_environment`?
Again Typst integration is not stable, as it is only available inside the pre-release and as stated several Quarto features are not available, so are some Typst CLI features.

Also without any reprex to have a setup close to yours, it's hard to help.

**Reply by mcanouil** [reply]
I was able to create something, currently Quarto does not ingest the command nor the environment variable and I think it should (cc @cderv).

Current workaround is to use `keep-typ`, then `quarto typst compile your-document.typ`.

MRE: https://github.com/mcanouil/quarto-issues/tree/main/quarto-cli-7580

**Reply by royfrancis** [reply]
I tried `_environment` after your suggestion. It didn't help. I tried relative and absolute paths to the fonts directory. Didn't see a difference. Then again, I am not in a project. I just have one qmd file. So, `_environment` is probably not even used.

**Reply by royfrancis** [reply]
That worked! So first `quarto render test.qmd` with 

```
format:
  typst:
    keep-typ: true
```

And then 

`quarto typst compile test.typ --font-path ./fonts/`

**Reply by cderv** [reply]
Thanks for looking into this. 

> Current workaround is to use keep-typ, then quarto typst compile your-document.typ.

This works because you are bypassing Quarto rendering using typst. 

When quarto calls typst, we do set `--font-path` already at command line to add some fonts required for callouts. It seems that typst, does not allow setting an env var + a CLI flag. So our call will override anything set in `TYPST_FONT_PATHS`

Try for example on linux 

````bash
export TYPST_FONT_PATHS="./fonts" # or $env:TYPST_FONT_PATHS="./fonts" on PowerShell
quarto typst fonts --font-path ./dummy
````

The `--font-path` will take precedence over the env var. Trying 

````bash
export TYPST_FONT_PATHS="./fonts" # or $env:TYPST_FONT_PATHS="./fonts" on PowerShell
quarto typst fonts
````

will now show the font. 

I am surprised by this behavior, so I'll look closer on how typst works, and what we can do. We may need to merge user fonts path with our internal font paths.

**Reply by seawaR** [reply]
Hey @royfrancis, I was wondering if you found a way to use a custom font in a template without the workaround.

**Comment by mcanouil** [comment, marked as answer]
Closing in favour of:
- https://github.com/quarto-dev/quarto-cli/issues/7585

**Reply by mcanouil** [reply]
@seawaR See the closed issue.
